### PR TITLE
feat: Datastore paging, sorting, filtering [DHIS2-12329]

### PR DESCRIPTION
**Pull request by jbee**
### Summary
Provides paging, filtering and sort order for datastore.

Few specials over how metadata works:

* `_` as property name in filters can be used to mean the datastore entry key
* `.` as property name in filters can be used to mean the root JSON value
* sort order supports 4 types
  * `asc`: alphabetically ascending (15 is before 2)
  * `desc`: alphabetically descending
  * `nasc`: numeric ascending (2 is before 15; fails if values aren't numbers)
  * `ndesc`: numeric descending (fails if values aren't numbers)
* `null` as filter value will only look for JSON nodes of type _null_ (unless a text operator is used)
* `true`/`false` as filter value will only look for JSON nodes of type _boolean_ (unless a text operator is used)
* numeric value for filters will only look for JSON nodes of type _number_ (unless a text operator is used)
* a value can be single quoted to force interpretation as a _string_, e.g. `'10'` or `'true'`
* `empty` operator matches empty objects, arrays and string values

### Spec
For a more detailed spec of the feature see https://github.com/dhis2/notes/discussions/321

### JSONB Functions
An overview of what jsonb functions are used

* `jsonb_extract_path` (essential)
* `jsonb_extract_path_text` (essential for filters)
* `jsonb_typeof` (essential for filters)
* `to_jsonb` (many filters)
* `jsonb_array_length` (emptiness filter only)
* `jsonb_object` (emptiness filter only)

### Changes
Firstly this is a much smaller actual change than it looks from the files changed. Vast majority is just touched because of the renaming I made (with Lars' blessing - see Rename)

Actual changes are in:
* `ErrorCode` (added code)
* `DatastoreQueryValidator` (new file to valid query consistency)
* `DatastoreQueryBuilder` (new file to build HQL)
* `HibernateDatastoreStore` (use builder)
* `DatastoreController` (minimal extension forwarding parameters to build query)
* as well as new tests scenarios and classes

**Renames:**
* the `*KeyJsonValue*` pattern became the `*Datastore*` pattern
* methods and parameters with `KeyJsonValue` in the name usually now only use `Entry` instead

### Automatic Testing
Added tests for `fields` extraction, `order` sorting and `filter` filtering. However, due to limitations with H2 database the filters cannot cover test scenarios where the filtered property is the JSON value. All filter types and features have been tested with the entry key as the filter property. This tests most of the query building code and the part that would apply these to the JSON value can be tested when testing field extraction. The issue here is that H2 the extracted type is not suitable for the comparison (which is incorrect) but it can be used to sort or extract.

### Manual Testing
As filters on value could not be covered with automatic tests these should be tested extensively using manual tests.

Scenarios are `filter=...`:
*  `<property>:null`
* `<property>:!null`
* `<property>:empty`
* `<property>:!empty`
* `<property>:eq:<value>`
* `<property>:!eq:<value>`
* `<property>:ieq:<value>`
* `<property>:in:[<value>,<value>,...]`
* `<property>:!in:[<value>,<value>,...]`
* `<property>:lt:<value>`
* `<property>:le:<value>`
* `<property>:gt:<value>`
* `<property>:ge:<value>`
* `<property>:like:<value>`
* `<property>:!like:<value>`
* ... (all the other pattern matching variants)

Each of these should be tested once with the root value property `.` and with a path property, something like `name`, and a nested path property, something like `parent.name`.
Also these should be tested for the entry key `_` property.
In addition equality needs testing for text, number and boolean values, again on root and path.
Relational inequality needs testing for text and number values, again on root and path.

In order to allow for meaningful testing it is necessary to first setup a pool of 10ish entries in a test namespace that has suitable JSON values. About halve with simple JSON values, halve with objects and arrays.

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/datastore/DatastoreEntry.java [hunk 1]
```diff
@@ -45,7 +45,7 @@
 @ToString( onlyExplicitlyIncluded = true )
 @EqualsAndHashCode( onlyExplicitlyIncluded = true, callSuper = true )
 @NoArgsConstructor
-public class KeyJsonValue extends BaseIdentifiableObject
+public class DatastoreEntry extends BaseIdentifiableObject
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/datastore/DatastoreEntry.java:48]: Should we call this `DataStoreEntry` with capital S? The internet seems to think that splitting data store on two words is most common.

**Comment by larshelge** [comment]
Would be nice to have a few integration tests against PostgreSQL using testcontainers.

**Comment by larshelge** [comment]
PR has one conflict against master now for JsonKeyJsonValue.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/datastore/DatastoreEntry.java (hunk 1 above)
Comment (jbee) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/datastore/DatastoreEntry.java:48]: Have no strong opinion. My thought and argument making it one word is that it has less potential confusing this with other things that have `Data` as a prefix. `Datastore` as a prefix is less ambiguous or more specific.

**Review by larshelge — APPROVED** [review]

**Review by vietnguyen — APPROVED** [review]

**Pull request merged**